pkg/operation/care/checker.go

```diff
@@ -257,13 +260,27 @@ func resourcesNotProgressingCheck() func(status gardencorev1beta1.ConditionStatu
 	}
 }
 
-func (b *HealthChecker) checkManagedResourceConditions(condition gardencorev1beta1.Condition, mr *resourcesv1alpha1.ManagedResource, conditionsToCheck map[gardencorev1beta1.ConditionType]func(status gardencorev1beta1.ConditionStatus) bool) *gardencorev1beta1.Condition {
+func (b *HealthChecker) checkManagedResourceConditions(
+	condition gardencorev1beta1.Condition,
+	mr *resourcesv1alpha1.ManagedResource,
+	conditionsToCheck map[gardencorev1beta1.ConditionType]func(status gardencorev1beta1.ConditionStatus) bool,
+	managedResourceProgressingThreshold *metav1.Duration) *gardencorev1beta1.Condition {
 	if mr.Generation != mr.Status.ObservedGeneration {
 		c := b.FailedCondition(condition, gardencorev1beta1.OutdatedStatusError, fmt.Sprintf("observed generation of managed resource '%s/%s' outdated (%d/%d)", mr.Namespace, mr.Name, mr.Status.ObservedGeneration, mr.Generation))
 		return &c
 	}
 
+	var conditionProgressing, conditionHealthy bool
 	for _, cond := range mr.Status.Conditions {
+		if cond.Type == resourcesv1alpha1.ResourcesProgressing &&
+			cond.Status == gardencorev1beta1.ConditionTrue &&
+			managedResourceProgressingThreshold != nil &&
+			time.Since(cond.LastTransitionTime.Time) >= managedResourceProgressingThreshold.Duration {
+			conditionProgressing = true
+		}
+		if cond.Type == resourcesv1alpha1.ResourcesHealthy && cond.Status == gardencorev1beta1.ConditionTrue {
+			conditionHealthy = true
+		}
```

Comment:
Is it possible to perform this check via the `conditionsToCheck` map that is passed? I see that currently the func in `conditionsToCheck` accepts only the condition status but we could give a try to adapt it to accept the whole ManagedResource?
I am writing down this comment because `ResourcesApplied` and `ResourcesHealthy` conditions are checked via funcs in this `conditionsToCheck` map but on the other side we have custom coding inline for the Progressing condition.

Reply:
I think keeping it this way is better -> This is not just checking `ResourceProgressing` condition but depends on `ResourcesHealthy` so to make it work via `conditionsToCheck ` whole managed resource needs to be passed in the function, and this is not the main reason. The main I prefer not to do so is - 
For all conditions, it will be required to loop through all conditions so that 3*3 total 9 operations but in the current scenario, it is required only 3. So if there are many shoots in one seed this can created overhead.
But if it is not important I can change it. WDYT?

Reply:
It is more challenging to perform the check via the `conditionsToCheck` map.

I don't think it is that important how many times (3 or 9) we will loop over the conditions. I could only came with the following suggestion about improving the readability:

```diff
diff --git a/pkg/operation/care/checker.go b/pkg/operation/care/checker.go
index 5166968cb..b2d04a934 100644
--- a/pkg/operation/care/checker.go
+++ b/pkg/operation/care/checker.go
@@ -271,17 +271,7 @@ func (b *HealthChecker) checkManagedResourceConditions(
 		return &c
 	}

-	var conditionProgressingTooLong, conditionHealthy bool
 	for _, cond := range mr.Status.Conditions {
-		if cond.Type == resourcesv1alpha1.ResourcesProgressing &&
-			cond.Status == gardencorev1beta1.ConditionTrue &&
-			b.managedResourceProgressingThreshold != nil &&
-			b.clock.Since(cond.LastTransitionTime.Time) >= b.managedResourceProgressingThreshold.Duration {
-			conditionProgressingTooLong = true
-		}
-		if cond.Type == resourcesv1alpha1.ResourcesHealthy && cond.Status == gardencorev1beta1.ConditionTrue {
-			conditionHealthy = true
-		}
 		checkConditionStatus, ok := conditionsToCheck[cond.Type]
 		if !ok {
 			continue
@@ -302,6 +292,14 @@ func (b *HealthChecker) checkManagedResourceConditions(
 		return &c
 	}

+	resourcesHealthyCondition := v1beta1helper.GetCondition(mr.Status.Conditions, resourcesv1alpha1.ResourcesHealthy)
+	resourcesProgressingCondition := v1beta1helper.GetCondition(mr.Status.Conditions, resourcesv1alpha1.ResourcesProgressing)
+	conditionHealthy := resourcesHealthyCondition != nil && resourcesHealthyCondition.Status == gardencorev1beta1.ConditionTrue
+	conditionProgressingTooLong := resourcesProgressingCondition != nil &&
+		resourcesProgressingCondition.Status == gardencorev1beta1.ConditionTrue &&
+		b.managedResourceProgressingThreshold != nil &&
+		b.clock.Since(resourcesProgressingCondition.LastTransitionTime.Time) >= b.managedResourceProgressingThreshold.Duration
+
 	if conditionProgressingTooLong && conditionHealthy {
 		c := b.FailedCondition(condition, gardencorev1beta1.ManagedResourceProgressingRolloutStuck, fmt.Sprintf("ManagedResource %s is progressing for more than %s", mr.Name, b.managedResourceProgressingThreshold.Duration))
 		return &c
```

It keeps the first look short and focused only on the checks from the `conditionsToCheck` map. The newly introduced check is moved afterwards.

Reply:
Draft approach using the `conditionsToCheck` map.

```diff
diff --git a/pkg/operation/care/checker.go b/pkg/operation/care/checker.go
index 5166968cb..7296fdcc5 100644
--- a/pkg/operation/care/checker.go
+++ b/pkg/operation/care/checker.go
@@ -241,52 +241,51 @@ func (b *HealthChecker) checkNodes(condition gardencorev1beta1.Condition, nodes
 
 // CheckManagedResource checks the conditions of the given managed resource and reflects the state in the returned condition.
 func (b *HealthChecker) CheckManagedResource(condition gardencorev1beta1.Condition, mr *resourcesv1alpha1.ManagedResource) *gardencorev1beta1.Condition {
-	conditionsToCheck := map[gardencorev1beta1.ConditionType]func(status gardencorev1beta1.ConditionStatus) bool{
-		resourcesv1alpha1.ResourcesApplied: defaultSuccessfulCheck(),
-		resourcesv1alpha1.ResourcesHealthy: defaultSuccessfulCheck(),
+	conditionsToCheck := map[gardencorev1beta1.ConditionType]func(condition gardencorev1beta1.Condition) bool{
+		resourcesv1alpha1.ResourcesApplied:     defaultSuccessfulCheck(),
+		resourcesv1alpha1.ResourcesHealthy:     defaultSuccessfulCheck(),
+		resourcesv1alpha1.ResourcesProgressing: resourcesNotProgressingCheck(b.clock, b.managedResourceProgressingThreshold),
 	}
 
 	return b.checkManagedResourceConditions(condition, mr, conditionsToCheck)
 }
 
-func defaultSuccessfulCheck() func(status gardencorev1beta1.ConditionStatus) bool {
-	return func(status gardencorev1beta1.ConditionStatus) bool {
-		return status != gardencorev1beta1.ConditionFalse && status != gardencorev1beta1.ConditionUnknown
+func defaultSuccessfulCheck() func(condition gardencorev1beta1.Condition) bool {
+	return func(condition gardencorev1beta1.Condition) bool {
+		return condition.Status != gardencorev1beta1.ConditionFalse && condition.Status != gardencorev1beta1.ConditionUnknown
 	}
 }
 
-func resourcesNotProgressingCheck() func(status gardencorev1beta1.ConditionStatus) bool {
-	return func(status gardencorev1beta1.ConditionStatus) bool {
-		return status != gardencorev1beta1.ConditionTrue && status != gardencorev1beta1.ConditionUnknown
+func resourcesNotProgressingCheck(clock clock.Clock, theshold *metav1.Duration) func(gardencorev1beta1.Condition) bool {
+	return func(condition gardencorev1beta1.Condition) bool {
+		notProgressing := condition.Status != gardencorev1beta1.ConditionTrue && condition.Status != gardencorev1beta1.ConditionUnknown
+
+		if theshold != nil && !notProgressing && clock.Since(condition.LastTransitionTime.Time) < theshold.Duration {
+			// ManagedResource is progressing but the given threshold didn't pass.
+			// Hence, return that the ManagedResource is not progressing.
+			return true
+		}
+
+		return notProgressing
 	}
 }
 
 func (b *HealthChecker) checkManagedResourceConditions(
 	condition gardencorev1beta1.Condition,
 	mr *resourcesv1alpha1.ManagedResource,
-	conditionsToCheck map[gardencorev1beta1.ConditionType]func(status gardencorev1beta1.ConditionStatus) bool,
+	conditionsToCheck map[gardencorev1beta1.ConditionType]func(condition gardencorev1beta1.Condition) bool,
 ) *gardencorev1beta1.Condition {
 	if mr.Generation != mr.Status.ObservedGeneration {
 		c := b.FailedCondition(condition, gardencorev1beta1.OutdatedStatusError, fmt.Sprintf("observed generation of managed resource '%s/%s' outdated (%d/%d)", mr.Namespace, mr.Name, mr.Status.ObservedGeneration, mr.Generation))
 		return &c
 	}
 
-	var conditionProgressingTooLong, conditionHealthy bool
 	for _, cond := range mr.Status.Conditions {
-		if cond.Type == resourcesv1alpha1.ResourcesProgressing &&
-			cond.Status == gardencorev1beta1.ConditionTrue &&
-			b.managedResourceProgressingThreshold != nil &&
-			b.clock.Since(cond.LastTransitionTime.Time) >= b.managedResourceProgressingThreshold.Duration {
-			conditionProgressingTooLong = true
-		}
-		if cond.Type == resourcesv1alpha1.ResourcesHealthy && cond.Status == gardencorev1beta1.ConditionTrue {
-			conditionHealthy = true
-		}
 		checkConditionStatus, ok := conditionsToCheck[cond.Type]
 		if !ok {
 			continue
 		}
-		if !checkConditionStatus(cond.Status) {
+		if !checkConditionStatus(cond) {
 			c := b.FailedCondition(condition, cond.Reason, cond.Message)
 			return &c
 		}
@@ -302,11 +301,6 @@ func (b *HealthChecker) checkManagedResourceConditions(
 		return &c
 	}
 
-	if conditionProgressingTooLong && conditionHealthy {
-		c := b.FailedCondition(condition, gardencorev1beta1.ManagedResourceProgressingRolloutStuck, fmt.Sprintf("ManagedResource %s is progressing for more than %s", mr.Name, b.managedResourceProgressingThreshold.Duration))
-		return &c
-	}
-
 	return nil
 }
 
diff --git a/pkg/operation/care/seed_health.go b/pkg/operation/care/seed_health.go
index 7752279d2..9232415f0 100644
--- a/pkg/operation/care/seed_health.go
+++ b/pkg/operation/care/seed_health.go
@@ -144,10 +144,10 @@ func (h *SeedHealth) checkSeedSystemComponents(
 }
 
 func checkManagedResourceForSeed(checker *HealthChecker, condition gardencorev1beta1.Condition, managedResource *resourcesv1alpha1.ManagedResource) *gardencorev1beta1.Condition {
-	conditionsToCheck := map[gardencorev1beta1.ConditionType]func(status gardencorev1beta1.ConditionStatus) bool{
+	conditionsToCheck := map[gardencorev1beta1.ConditionType]func(condition gardencorev1beta1.Condition) bool{
 		resourcesv1alpha1.ResourcesApplied:     defaultSuccessfulCheck(),
 		resourcesv1alpha1.ResourcesHealthy:     defaultSuccessfulCheck(),
-		resourcesv1alpha1.ResourcesProgressing: resourcesNotProgressingCheck(),
+		resourcesv1alpha1.ResourcesProgressing: resourcesNotProgressingCheck(checker.clock, nil),
 	}
 
 	return checker.checkManagedResourceConditions(condition, managedResource, conditionsToCheck)

```

WDYT?

Reply:
It works, I made the changes in 54eee174f8eecd8e372978edef01e3fe06c58305. My initial idea was to check both `Progressing` and `Healthy` but checking only the `Progressing` condition makes things simpler.